CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc

```diff
@@ -86,14 +104,29 @@ std::pair<int, PrimaryVertexAssignment::Quality> PrimaryVertexAssignment::charge
     }
   }
 
-  // first use "closest in Z" with tight cuts (targetting primary particles)
-  const float add_cov = vtxIdMinSignif >= 0 ? vertices[vtxIdMinSignif].covariance(2, 2) : 0.f;
-  const float dzE = sqrt(track->dzError() * track->dzError() + add_cov);
-  if (vtxIdMinSignif >= 0 and
-      (dzmin < maxDzForPrimaryAssignment_ and dzmin / dzE < maxDzSigForPrimaryAssignment_ and
-       track->dzError() < maxDzErrorForPrimaryAssignment_) and
-      (!useTime or dtmin / timeReso < maxDtSigForPrimaryAssignment_)) {
-    iVertex = vtxIdMinSignif;
+  // protect high pT particles from association to pileup vertices and assign them to the first vertex
+  if ((fPtMaxCharged_ > 0) && (vtxIdMinSignif >= 0) && (track->pt() > fPtMaxCharged_)) {
+    iVertex = 0;
+  } else {
+    // first use "closest in Z" with tight cuts (targetting primary particles)
+    const float add_cov = vtxIdMinSignif >= 0 ? vertices[vtxIdMinSignif].covariance(2, 2) : 0.f;
+    const float dzE = sqrt(track->dzError() * track->dzError() + add_cov);
+    if (!fOnlyUseFirstDz_) {
+      if (vtxIdMinSignif >= 0 and
+          (dzmin < maxDzForPrimaryAssignment_ and dzmin / dzE < maxDzSigForPrimaryAssignment_ and
+           track->dzError() < maxDzErrorForPrimaryAssignment_) and
+          (!useTime or dtmin / timeReso < maxDtSigForPrimaryAssignment_))
+        iVertex = vtxIdMinSignif;
+    } else {
+      // consider only distances to first vertex for association of pileup vertices (originally used in PUPPI)
+      if ((vtxIdMinSignif >= 0) && (std::abs(track->eta()) > fEtaMinUseDz_))
+        iVertex = ((std::abs(track->dz(vertices.at(0).position())) < maxDzForPrimaryAssignment_ and
```

Comment:
`std::abs(track->dz(vertices.at(0).position()))` seems to be recomputed several times in this function. Perhaps it makes sense to assign it to a variable once?